proxy/http2/Http2ClientSession.cc

```diff
@@ -366,7 +366,10 @@ Http2ClientSession::main_event_handler(int event, void *edata)
 
   recursion--;
   if (!connection_state.is_recursing() && this->recursion == 0 && kill_me) {
-    this->free();
+    MUTEX_TRY_LOCK(lock, this->mutex, this_ethread());
+    if (lock.is_locked()) {
+      this->free();
+    }
```

Comment:
Are we concerned with leaking?

Also at the top of the handler we have an assert that the mutex is held by the current thread.

In the master version of Continuation::handleEvent, there is an ink_release_assert that the mutex is held.  So the additional lock attempt here seems redundant.

Reply:
> Are we concerned with leaking?

Probably the thread have the lock will go to A in above. Otherwise, `TS_HTTP_SSN_CLOSE_HOOK` doesn't work. But it's possible.

> Also at the top of the handler we have an assert that the mutex is held by the current thread.

It's `ink_assert` not `ink_release_assert` unfortunately.

> In the master version of Continuation::handleEvent, there is an ink_release_assert that the mutex is held. So the additional lock attempt here seems redundant.

I agree. It's redundant on master. But we don't have it on 8.0.x or 7.1.x yet, right? 
Should we backport them instead of this?

Reply:
@masaori335  Why are we running into this now?  Why is another thread freeing client session?

Reply:
Talked on IRC, Http2Stream(s) stay on the same thread which created it. OTOH, it's possible that Http2ClientSession/Http2ConnectionState jump threads. Probably, it's root cause. From performance point of view, this should be fixed.

Reply:
I think it might be good to bring the ink_release_assert in Continuation::handleEvent over to 8.0 at least.  Better to get to the bottom of the race rather than patch around it.